### PR TITLE
Refactor sync::Once

### PR DESCRIPTION
**Pull request by pitdicker**
`std::sync::Once` contains some tricky code to park and unpark waiting threads. [once_cell](https://github.com/matklad/once_cell) has very similar code copied from here. I tried to add more comments and refactor the code to make it more readable (at least in my opinion). My PR to `once_cell` was rejected, because it is an advantage to remain close to the implementation in std, and because I made a mess of the atomic orderings. So now a PR here, with similar changes to `std::sync::Once`!

The initial goal was to see if there is some way to detect reentrant initialization instead of deadlocking. No luck there yet, but you first have to understand and document the complexities of the existing code :smile:.

*Maybe not this entire PR will be acceptable, but I hope at least some of the commits can be useful.*

Individual commits:

#### Rename state to state_and_queue
Just a more accurate description, although a bit wordy. It helped me on a first read through the code, where before `state` was use to encode a pointer in to nodes of a linked list.

#### Simplify loop conditions in RUNNING and add comments
In the relevant loop there are two things to be careful about:
- make sure not to enqueue the current thread only while still RUNNING, otherwise we will never be woken up (the status may have changed while trying to enqueue this thread).
- pick up if another thread just replaced the head of the linked list.

Because the first check was part of the condition of the while loop, the rest of the parking code also had to live in that loop. It took me a while to get the subtlety here, and it should now be clearer.

Also call out that we really have to wait until signaled, otherwise we leave a dangling reference.

#### Don't mutate waiter nodes
Previously while waking up other threads the managing thread would `take()` out the `Thread` struct and use that to unpark the other thread. It is just as easy to clone it, just 24 bytes. This way `Waiter.thread` does not need an `Option`, `Waiter.next` does not need to be a mutable pointer, and there is less data that needs to be synchronised by later atomic operations.

#### Turn Finish into WaiterQueue
In my opinion these changes make it just a bit more clear what is going on with the thread parking stuff.

#### Move thread parking to a seperate function
Maybe controversial, but with this last commit all the thread parking stuff has a reasonably clean seperation from the state changes in `Once`. This is arguably the trickier part of `Once`, compared to the loop in `call_inner`. It may make it easier to reuse parts of this code (see https://github.com/rust-lang/rfcs/pull/2788#discussion_r336729695). Not sure if that ever becomes a reality though.

#### Reduce the amount of comments in call_inner
With the changes from the previous commits, the code pretty much speaks for itself, and the amount of comments is hurting readability a bit.

#### Use more precise atomic orderings
Now the hard one. This is the one change that is not anything but a pure refactor or change of comments.

I have a dislike for using `SeqCst` everywhere, because it hides what the atomics are supposed to do. the rationale was:
> This cold path uses SeqCst consistently because the performance difference really does not matter there, and SeqCst minimizes the chances of something going wrong.

But in my opinion, having the proper orderings and some explanation helps to understand what is going on. My rationale for the used orderings (also included as comment):

When running `Once` we deal with multiple atomics: `Once.state_and_queue` and an unknown number of `Waiter.signaled`.
* `state_and_queue` is used (1) as a state flag, (2) for synchronizing the data that is the result of the `Once`, and (3) for synchronizing `Waiter` nodes.
    - At the end of the `call_inner` function we have to make sure the result of the `Once` is acquired. So every load which can be the only one to load COMPLETED must have at least Acquire ordering, which means all three of them.
    - `WaiterQueue::Drop` is the only place that may store COMPLETED, and must do so with Release ordering to make result available.
    - `wait` inserts `Waiter` nodes as a pointer in `state_and_queue`, and needs to make the nodes available with Release ordering. The load in its `compare_and_swap` can be Relaxed because it only has to compare the atomic, not to read other data.
    - `WaiterQueue::Drop` must see the `Waiter` nodes, so it must load `state_and_queue` with Acquire ordering.
    - There is just one store where `state_and_queue` is used only as a state flag, without having to synchronize data: switching the state from INCOMPLETE to RUNNING in `call_inner`. This store can be Relaxed, but the read has to be Acquire because of the requirements mentioned above.
* `Waiter.signaled` is both used as a flag, and to protect a field with interior mutability in `Waiter`. `Waiter.thread` is changed in `WaiterQueue::Drop` which then sets `signaled` with Release ordering. After `wait` loads `signaled` with Acquire and sees it is true, it needs to see the changes to drop the `Waiter` struct correctly.
* There is one place where the two atomics `Once.state_and_queue` and `Waiter.signaled` come together, and might be reordered by the compiler or processor. Because both use Aquire ordering such a reordering is not allowed, so no need for SeqCst.

cc @matklad

### CODE REVIEW

**Comment by pitdicker** [comment]
Thinking a bit more about this, I don't think the commit 'Don't mutate waiter nodes' is the smartest move. Let me work on it a little more before anyone takes a look to review (will probably take me a day).

**Comment by pitdicker** [comment]
Added another commit.

#### In Waiter use interior mutability and use no raw pointer

The existing code first created a mutable reference to a `Waiter` struct, turned it into a raw mutable pointer, and passed it to another thread. Then it kept using a reference to the atomic `signaled` inside that `Waiter`. If I understand things correctly that is unsound, but this is at the edge of my knowledge.

Two options we can choose from instead:
- Pass a raw non-mutable pointer to the other thread. In `WaiterQueue::drop` make a clone from the `thread` and `next` fields. This seemed suboptimal to me: cloning `thread` means cloning and dropping an `Arc`; and copying `next` can leave a dangling pointer around in the `Waiter` struct.
- Pass a shared reference to the other thread. Use interior mutability, where `WaiterQueue::drop` swaps out the `thread` and `next` fields with `None`.

This commit goes the route with interior mutability, and minimizing the use of raw pointers.

cc @ralfjung You may be interested in what I think are two cases of unsoundness in `Once`:
- Handing out a `*mut Waiter` to `node` [here](https://github.com/rust-lang/rust/pull/65719/files#diff-80754b8db8699947d7b2a43a9cc17dedL408) and later doing a `node.signaled.load` [here](https://github.com/rust-lang/rust/pull/65719/files#diff-80754b8db8699947d7b2a43a9cc17dedL424).
- What may be another case of https://github.com/rust-lang/rust/issues/55005 [here](https://github.com/rust-lang/rust/pull/65719/files#diff-80754b8db8699947d7b2a43a9cc17dedR519-R522).

**Comment by oliver-giersch** [comment]
From what I can see, this is not only a refactoring, but also a fix for UB (aliasing of a mutable reference), very good catch!
I haven't looked through the code in its entirety, but some things stuck out:
- I don't believe you need interior mutability for the stack waiter's next pointer/reference, since it is only updated during the CAS loop in the failure case, i.e. before the waiter becomes visible to other threads
- I don't what the `RUNNING` case is necessary for (in the original implementation as well), any bit pattern that is not 0, 1, 2 or 3 (for the various states) is necessarily a pointer to a waiter, so you could do away with this state and the bitwise and required to extract the pointer
- memory orderings are notoriously difficult to get correct ofc, but in my estimation the implementation does not require `SeqCst` at all

I have updated my own implementation of [`conquer_once`](https://github.com/oliver-giersch/conquer-once) to fix the UB and also added some comments explaining my reasoning on the various memory orderings. If you like, you can cross-reference and compare with your own assessments.

**Comment by pitdicker** [comment]
> * I don't believe you need interior mutability for the stack waiter's next pointer/reference, since it is only updated during the CAS loop in the failure case, i.e. before the waiter becomes visible to other threads

Yes, that is possible. I choose to also switch it back to `None` here when threads get woken up. This way there is no dangling pointer from one woken up thread to another woken thread, and we can use normal references. Seemed just a bit nicer to me, but doesn't really matter.

> * I don't what the `RUNNING` case is necessary for (in the original implementation as well), any bit pattern that is not 0, 1, 2 or 3 (for the various states) is necessarily a pointer to a waiter, so you could do away with this state and the bitwise and required to extract the pointer

You probably also considered the case where there are no waiting threads and the pointer part is 0. If I understand it right you propose that both RUNNING (0x2) and any pattern not matching 0, 1, 2, or 3 correspond to RUNNING? That could work, but I don't yet see an advantage to warrant the subtlety. 

> * memory orderings are notoriously difficult to get correct ofc, but in my estimation the implementation does not require `SeqCst` at all

Then we mostly agree. What did you think about the argument that `state.load` may be reordered by the processor with `signaled.load`? I see it as very implausible, but would want to use `SeqCst` just be be safe for these two.

**Comment by pitdicker** [comment]
Your trick to use `#[repr(align(4))]` on `Waiter` in `conquer-once` is a good idea, that will guarantee we have 2 bits for state, even if the platform might want to align differently.

Also using the names `StackWaiter` and `Stackwaiter.ready` seems nice.

**Comment by oliver-giersch** [comment]
> Yes, that is possible. I choose to also switch it back to `None` here when threads get woken up. This way there is no dangling pointer from one woken up thread to another woken thread, and we can use normal references. Seemed just a bit nicer to me, but doesn't really matter.

I would also argue that using a raw `*const` or perhaps a `NonNull` for the next pointer would be the cleaner and also safer way, than to store and dereference to references with arbitrary lifetime. The lifetime does not have any meaning, given that the references point into the stacks of *other* blocked threads .
Also, if dangling references are really a problem/UB (which I am not necessarily convinced of, yet), then [this](https://github.com/rust-lang/rust/blob/d9996cc35d9681794661163390cac3edfe46ef55/src/libstd/sync/once.rs#L505) might also be problematic, I think.
```rust
let mut curr: *const Waiter = ...;
while !curr.is_null() {
    let thread = unsafe {
        let next = (*curr).next;
        let thread = (*curr).thread.take().unwrap();
        (*curr).signaled.store(true, Release);
        thread
    }
    thread.unpark();
}
```

looks cleaner to me, alternatively you could also create a temporary reference inside the unsafe block, of course.

> You probably also considered the case where there are no waiting threads and the pointer part is 0. If I understand it right you propose that both RUNNING (0x2) and any pattern not matching 0, 1, 2, or 3 correspond to RUNNING? That could work, but I don't yet see an advantage to warrant the subtlety.

Sorry, I wrote this from memory and have confused some things.
In my implementation, the initial state is 1 and the first accessing thread sets it to 0 (or `null`) and all further threads interpret it as queue and insert their `Waiter` pointers. It's a minor thing, but it means you don't need bitwise logic  and the `STATE_MASK` constant for enqueuing or dequeueing waiters.

> Then we mostly agree. What did you think about the argument that `state.load` may be reordered by the processor with `signaled.load`? I see it as very implausible, but would want to use `SeqCst` just be be safe for these two.

IIRC the memory model does not permit the reordering of non-relaxed atomic operations among themselves, but I'd have to find a reference for that.

**Comment by pitdicker** [comment]
> I would also argue that using a raw `*const` or perhaps a `NonNull` for the next pointer would be the cleaner and also safer way, than to store and dereference to references with arbitrary lifetime. The lifetime does not have any meaning, given that the references point into the stacks of _other_ blocked threads.

I am not married to the idea, but wanted to give it a try. Not smart of me, as I also wanted this PR to be mostly uncontroversial. I do believe the lifetimes work out, but will just as easily revert that part if other reviewers feel the same way.

**Comment by pitdicker** [comment]
> IIRC the memory model does not permit the reordering of non-relaxed atomic operations among themselves, but I'd have to find a reference for that.

I was pointed by @matklad to https://internals.rust-lang.org/t/idea-replace-seqcst-with-seqcst-acq-rel-acqrel/11028/3

**Comment by pitdicker** [comment]
@oliver-giersch I have reverted the part about using normal references, now things are back to raw pointers as at the start of this PR. And I have changed the stuff about how two loads might need to be SeqCst.
The commits are rebased, and the first post is partly updated to keep things a bit sane.

**Comment by pitdicker** [comment]
> One question though, what was your reasoning for going with acquire-release in [this](https://github.com/rust-lang/rust/blob/c2bbfeadcce08a4b8ce02b66906ecc542cc9df39/src/libstd/sync/once.rs#L486) instead of release?

Because it has to acquire the nodes from the waiting threads before walking the queue to wake them up.

**Comment by oliver-giersch** [comment]
> Because it has to acquire the nodes from the waiting threads before walking the queue to wake them up.

Yeah I believe you are correct. It's good to these exercises from time to time.

**Comment by JohnCSimon** [comment]
Ping from triage.
@KodrAus can you please review this PR?
cc: @pitdicker @oliver-giersch 
Thank you!

**Comment by RalfJung** [comment]
> Handing out a *mut Waiter to node here and later doing a node.signaled.load here.

That looks dangerous indeed! The question is, what is that `me` pointer used for in the mean time? Looks like it is shared with other threads, which fields to they access? Once the line with `node.signaled` is executed, that takes away the permission of `me` to access the `signaled` field.

> What may be another case of #55005 here.

Unfortunately the line numbers changed, so I am not sure what that link pointed to.
Could you use a link to a commit on master, instead of one into your PR?

File: src/libstd/sync/once.rs [hunk 1]
```diff
@@ -347,122 +388,127 @@ impl Once {
...
                 }
             }
         }
     }
 }
 
+fn wait(state_and_queue: &AtomicUsize, current_state: usize) {
+    // Create the node for our current thread that we are going to try to slot
+    // in at the head of the linked list.
+    let mut node = Waiter {
```
Comment (RalfJung) [inline src/libstd/sync/once.rs]: The `node` is still mutable so you still have a problem, now with the `next` field.
You could try wrapping that in a `Cell`, too, and that should actually work as you never create a mutable reference or use mutable accesses that would assert uniqueness. If feel making sure all accesses go through a raw pointer would be easier though.

File: src/libstd/sync/once.rs [hunk 2]
```diff
@@ -347,122 +388,127 @@ impl Once {
...
+    // Try to slide in the node at the head of the linked list.
+    // Run in a loop where we make sure the status is still RUNNING, and that
+    // another thread did not just replace the head of the linked list.
+    let mut old_head_and_status = current_state;
+    loop {
+        if old_head_and_status & STATE_MASK != RUNNING {
+            return; // No need anymore to enqueue ourselves.
+        }
+
+        node.next = (old_head_and_status & !STATE_MASK) as *const Waiter;
```
Comment (RalfJung) [inline src/libstd/sync/once.rs]: This is where you still write to `node.next`, which is as bad as doing `&mut node.next` -- it says that this is unique, meaning that future reads through `me` become illegal.

Comment (pitdicker) [reply src/libstd/sync/once.rs]: You are right! I'll think about it a little, hopefully some reordering is also an alternative... And otherwise raw pointers all around, but that makes me less comfortable.

Comment (oliver-giersch) [reply src/libstd/sync/once.rs]: How can this be problematic? The mutable reference ist unique, since the next field is only written to *before* the node becomes visible to other threads. Isn't this like mutating some value and *then* sending a shared reference to some other thread?

Comment (RalfJung) [reply src/libstd/sync/once.rs]: You can't just take a random `usize` to the right address and use that; the corresponding address has to have an item in its borrow stack indicating that usize-ptrs are allowed to access it. This item gets added when you cast a reference to a raw pointer, but here it gets *removed* again!

When you use a mutable reference, that means "no other pointer is allowed to point here". This includes pointers "hidden" in integers. Maybe a better way to understand it is to think of this as: "all future accesses to this location must happen with pointers that are derived from me in the future". No pointer ever gets derived again from `node`, and thus no pointer ever again acquires the permission to access `node.next`.

**Comment by pitdicker** [comment]
I have moved stuff around so there is no more mutation to `node` happening inside `wait`. Would this be more up to scratch?

**Comment by RalfJung** [comment]
I can't at a glance judge the functional correctness of this, but `node` isn't mutable any more so there are likely no uniqueness assertions. There should be comments explaining why that's important and why we have `Cell`.

I'd personally still prefer raw pointers of interior mutability, but well, I won't fight the one who does the actual work here. ;)

**Comment by pitdicker** [comment]
> There should be comments explaining why that's important and why we have `Cell`.

There is something at https://github.com/rust-lang/rust/pull/65719/files#diff-80754b8db8699947d7b2a43a9cc17dedR169, but I'll add more.

> I'd personally still prefer raw pointers of interior mutability, but well, I won't fight the one who does the actual work here. ;)

Thank you! I don't mind which one is used, but I have some trouble picturing how the code should like like using raw pointers :-)

**Comment by pitdicker** [comment]
There is one more thing I was playing with, but didn't include until now.

For the `once_cell` crate Is was playing with the atomic Consume ordering. For that I wrote a test that catches insufficient synchronization pretty very when run on my phone. And to make sure there was not some other synchronization happening that hides shortcomings, I wanted to find a way to remove the acquire from the loop that checks `node.signaled`.

The acquire is currently necessary because the `node` is read shortly thereafter, when it gets dropped. But if it was able to successfully add itself to the queue, there isn't anything that needs dropping: another thread takes out `thread` and takes care of dropping that.

Now what if we wrap `node` inside `ManuallyDrop` (and are a little careful)? It seems to me we then no longer need to acquire `node` again after getting unparked, because we never read from it again.

Is that a correct conclusion, that we would not need to acquire data if we don't read it anymore? Even if it lives on our stack etc.? Is it true that using a `cell` prevents optimistic reads, if those are even a concern?

**Comment by RalfJung** [comment]
> For the once_cell crate Is was playing with the atomic Consume ordering

(Stable) Rust doesn't support "consume", and for a good reason -- nobody knows how to properly specify it. Let's just stick with the sane and well-understood release/acquire orderings.

**Comment by pitdicker** [comment]
I don't want to suggest using it here. But I was curious about if we wouldn't need to do an acquire if we don't do a read from the non-atomic fields of `node` after getting unparked.

**Comment by RalfJung** [comment]
That sentence contains too many negations for me to be able to parse it.^^

**Comment by pitdicker** [comment]
Let's forget about it for now. It was something I was wondering about, but not planning for this PR.

I have amended the last commit to keep the comments in `wait` helpful, with on top in `wait`:

> Note: the following code was carefully written to avoid creating a mutable reference to `node` that gets aliased.

File: src/libstd/sync/once.rs (hunk 2 above)
Comment (oliver-giersch) [reply src/libstd/sync/once.rs]: Wat, is any of this documented somewhere?

So `me` is a raw `*const` derived from a shared reference and instantly turned into a `usize`. Then it is mutated (using a `&mut`) once or a bunch of times and then another shared reference is created to read from the atomic `signalled` variable, wouldn't that mean that there *is* a shared reference in this borrow stack?

Comment (RalfJung) [reply src/libstd/sync/once.rs]: > Wat, is any of this documented somewhere?

The normative docs are "don't violate the aliasing rules". Whatever that means.

My proposed model is documented at https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md.

> So me is a raw *const derived from a shared reference and instantly turned into a usize. Then node it is mutated (using a &mut) once or a bunch of times and then another shared reference is created to read from the atomic signalled variable, wouldn't that mean that there is a shared reference in this borrow stack?

There's a shared reference on the borrow stack but no item permitting raw integers any access. Shared references are not a wildcard; they still carefully control that only currently valid shared references may be used. (See https://github.com/rust-lang/unsafe-code-guidelines/issues/87 for what happens when we are too relayed about shared references. It's not good.)

We have no clear idea yet what the interact between integer-ptr casts and provenance is. C and LLVM are suffering from [miscompilations](https://bugs.llvm.org/show_bug.cgi?id=35229) due to ignoring these issues.

**Comment by Amanieu** [comment]
The code LGTM. There's just a few formatting issues that a quick rustfmt can fix.

**Comment by pitdicker** [comment]
@Amanieu Thank you for reviewing. I have to admit this is my first time running rustfmt.
